src/H5Fsuper_cache.c

```diff
@@ -498,13 +512,22 @@ H5F__cache_superblock_deserialize(const void *_image, size_t H5_ATTR_NDEBUG_UNUS
             UINT16DECODE(image, chunk_btree_k);
 
             /* Reserved bytes are present only in version 1 */
-            if (sblock->super_vers == HDF5_SUPERBLOCK_VERSION_1)
+            if (sblock->super_vers == HDF5_SUPERBLOCK_VERSION_1) {
                 image += 2; /* reserved */
-        }                   /* end if */
+
+		/* Check whether the image pointer is out of bounds */
+		if (image >= image_end)
+		    HGOTO_ERROR(H5E_FILE, H5E_BADRANGE, NULL, "image pointer is out of bounds")
+	    }
+	}                   /* end if */
         else
             chunk_btree_k = HDF5_BTREE_CHUNK_IK_DEF;
         udata->btree_k[H5B_CHUNK_ID] = chunk_btree_k;
 
+	/* Check whether the image pointer will be out of bounds */
+	if (image + H5F_SIZEOF_ADDR(udata->f) * 4 >= image_end)
```

Comment:
Is there a reason this check looks ahead 4 times size of address when only 2 addresses are decoded? Same question for line 584.